### PR TITLE
[CI/meta] Add automatic labeling of changed files.

### PR DESCRIPTION
**Pull request by Kowlin**
### Type

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes

This is entering the never again bucket list! and HOPEFULLY! this should work, just a bit skeptical.

### CODE REVIEW

File: .github/workflows/auto_labeler.yml [hunk 1]
```diff
@@ -24,3 +24,7 @@ jobs:
               repo: context.repo.repo,
               labels: ['Status: Needs Triage']
             });
+      - name: Apply Type Label
+        uses: actions/labeler@v2
+        with:
+          repo-token: "${{ secrets.GITHUB_TOKEN }}"
```
Comment (Jackenmen) [inline .github/workflows/auto_labeler.yml]: This workflow does not run on new pull requests, only issues, so this won't work and it needs a separate workflow with `pull_request_target` event.

File: .github/labeler.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
+"Category: Admin":
+  - redbot/cogs/admin/*
+"Category: Alias":
+  - redbot/cogs/alias/*
+"Category: Audio API":
+  - redbot/cogs/audio/apis/*
+  - redbot/cogs/audio/data/*
+"Category: Audio Cog":
+  - redbot/cogs/audio/*
+  - redbot/cogs/audio/core/*
```
Comment (Jackenmen) [inline .github/labeler.yml]: Audio API is not yet a thing, so there's no actual folder related to it - everything in `redbot/cogs/audio` (except `locales` folder) should have the `Audio Cog` label applied
```suggestion
"Category: Audio Cog":
  - redbot/cogs/audio/**/*
  - !redbot/cogs/audio/**/locales/*
```

File: .github/labeler.yml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+"Category: Permissions":
+  - redbot/cogs/permissions/*
+"Category: Reports Cog":
+  - redbot/cogs/reports/*
+"Category: RPC/ZMQ API":
+  - redbot/core/rpc.py
+"Category: Streams":
+  - redbot/cogs/streams/*
+"Category: Tests":
+  - redbot/tests/*
```
Comment (Jackenmen) [inline .github/labeler.yml]: `tests/` folder is in main directory. There's also a `redbot/pytest` folder that keeps all the pytest fixtures.
```suggestion
  - redbot/pytest/*
  - tests/**/*
```

File: .github/labeler.yml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+"Category: Image":
+  - redbot/cogs/image/*
+"Category: Meta":
+  - ./*
+  - .github/*
+  - .github/ISSUE_TEMPLATE/*
+  - .github/PULL_REQUEST_TEMPLATE/*
+"Category: Mod Cog":
+  - redbot/cogs/mod/*
+"Category: Modlog API":
```
Comment (Jackenmen) [inline .github/labeler.yml:66]: `redbot/core/generic_casetypes.py`, `redbot/core/modlog.py`
```suggestion
"Category: Modlog API":
  - redbot/core/generic_casetypes.py
  - redbot/core/modlog.py
```

File: .github/labeler.yml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+"Category: Config":
+  - redbot/core/drivers/*
+  - redbot/core/config.py
+"Category: CustomCom":
+  - redbot/cogs/customcom/*
+"Category: Dev Cog":
+  - redbot/core/dev_commands.py
+"Category: Docs":
+  - docs/*
+  - docs/**/*
```
Comment (Jackenmen) [inline .github/labeler.yml]: This can be simplified:
```suggestion
"Category: Docs":
  - docs/**/*
```

File: .github/labeler.yml [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+"Category: i18n":
+  - redbot/core/i18n.py
+  - redbot/**/locales/*
+"Category: Image":
+  - redbot/cogs/image/*
+"Category: Meta":
+  - ./*
+  - .github/*
+  - .github/ISSUE_TEMPLATE/*
+  - .github/PULL_REQUEST_TEMPLATE/*
```
Comment (Jackenmen) [inline .github/labeler.yml:61]: `schema/*`, `tools/*`:
```suggestion
"Category: Meta":
  - ./*
  - .github/*
  - .github/ISSUE_TEMPLATE/*
  - .github/PULL_REQUEST_TEMPLATE/*
  - schema/*
  - tools/*
```

File: .github/labeler.yml [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  - redbot/core/_sharedlibdeprecation.py
+  - redbot/core/bot.py
+  - redbot/core/checks.py
+  - redbot/core/cli.py
+  - redbot/core/cog_manager.py
+  - redbot/core/core_commands.py
+  - redbot/core/errors.py
+  - redbot/core/events.py
+  - redbot/core/global_checks.py
+  - redbot/core/settings_caches.py
```
Comment (Jackenmen) [inline .github/labeler.yml]: `redbot/*`, `redbot/core/data_manager.py`:
```suggestion
"Category: Bot Core":
  - redbot/*
  - redbot/core/__init__.py
  - redbot/core/_sharedlibdeprecation.py
  - redbot/core/bot.py
  - redbot/core/checks.py
  - redbot/core/cli.py
  - redbot/core/cog_manager.py
  - redbot/core/core_commands.py
  - redbot/core/data_manager.py
  - redbot/core/errors.py
  - redbot/core/events.py
  - redbot/core/global_checks.py
  - redbot/core/settings_caches.py
```

**Review by Jackenmen — CHANGES_REQUESTED** [review]

**Comment by Jackenmen** [comment]
I also see that there's a `sync-labels` option in actions/labeler (defaults to false), which determines whether the labels should be removed when matching files are no longer changed by the PR. It might be worth considering?

And one question I haven't found an answer to... Will me removing labels from the PR manually is gonna turn into a fight once PR gets updated by the PR author? As in, will removing labels that shouldn't actually be applied end up with labeler reapplying them when PR gets updated by the author? As an example, #4371 would have a lot of labels applied and it's just a black reformat. I don't expect such situations to happen often, but I'm still wondering how that would work.

File: .github/workflows/auto_labeler.yml (hunk 1 above)
Comment (bobloy) [reply .github/workflows/auto_labeler.yml]: https://github.com/actions/labeler/issues/36
This issue documents jack's comment

**Review by Jackenmen — APPROVED** [review]

**Pull request merged**